third_party/proton/csrc/lib/Data/TraceData.cpp

```diff
@@ -251,44 +383,92 @@ void TraceData::dumpChromeTrace(std::ostream &os) const {
       uint64_t startTime =
           std::get<uint64_t>(kernelMetric->getValue(KernelMetric::StartTime));
       minTimeStamp = std::min(minTimeStamp, startTime);
+      hasKernelMetrics = true;
+    }
+    if (event.metrics.count(MetricKind::Cycle) && !hasKernelMetrics) {
+      auto context = trace->getContexts(event.contextId);
+      std::shared_ptr<CycleMetric> cycleMetric =
+          std::dynamic_pointer_cast<CycleMetric>(
+              event.metrics.at(MetricKind::Cycle));
+      auto uid = std::get<uint64_t>(cycleMetric->getValue(CycleMetric::UnitId));
+      auto startCycle =
+          std::get<uint64_t>(cycleMetric->getValue(CycleMetric::StartCycle));
+      auto endCycle =
+          std::get<uint64_t>(cycleMetric->getValue(CycleMetric::EndCycle));
+      auto blockId =
+          std::get<uint64_t>(cycleMetric->getValue(CycleMetric::BlockId));
+      auto procId =
+          std::get<uint64_t>(cycleMetric->getValue(CycleMetric::ProcessorId));
+      auto timeShiftCost =
+          std::get<uint64_t>(cycleMetric->getValue(CycleMetric::TimeShiftCost));
+
+      hasCycleMetrics = true;
+      // TODO(fywkevin): fix this kernrel name issue
+      const std::string &kernelName = context[1].name;
+      kernelBlockNum[kernelName] += 1;
+      cycleEvents.emplace_back(startCycle, endCycle, blockId, uid, procId,
+                               event.contextId, timeShiftCost);
     }
   }
 
-  // 3) for each streamId in ascending order, emit one JSON line
-  for (auto const &[streamId, events] : streamTraceEvents) {
-    json object = {{"displayTimeUnit", "us"}, {"traceEvents", json::array()}};
-
-    for (auto const &event : events) {
-      auto kernelMetrics = std::dynamic_pointer_cast<KernelMetric>(
-          event.metrics.at(MetricKind::Kernel));
-      uint64_t startTimeNs =
-          std::get<uint64_t>(kernelMetrics->getValue(KernelMetric::StartTime));
-      uint64_t endTimeNs =
-          std::get<uint64_t>(kernelMetrics->getValue(KernelMetric::EndTime));
-      uint64_t ts = startTimeNs - minTimeStamp;
-      uint64_t dur = endTimeNs - startTimeNs;
-
-      auto contextId = event.contextId;
-      auto contexts = trace->getContexts(contextId);
-
-      json element;
-      element["name"] = contexts.back().name;
-      element["cat"] = "kernel";
-      element["ph"] = "X";
-      element["ts"] = ts;
-      element["dur"] = dur;
-      element["tid"] = streamId; // thread id = stream
-      json callStack = json::array();
-      for (auto const &ctx : contexts) {
-        callStack.push_back(ctx.name);
-      }
-      element["args"]["call_stack"] = std::move(callStack);
+  assert(!(hasKernelMetrics && hasCycleMetrics) &&
```

Comment:
It could be a runtime error or unsupported error and you can put it in the while loop. In this way we don't need to write an end condition like `event.metrics.count(MetricKind::Kernel) && !hasCycleMetrics`. Just direct abort the execution when both are true.